File: haskell-server/src/BoardGame/Server/Domain/Board.hs
--
-- Copyright 2017 Azad Bolour
-- Licensed under GNU Affero General Public License v3.0 -
--   https://github.com/azadbolour/boardgame/blob/master/LICENSE.md
--

{-# LANGUAGE DeriveGeneric #-}
{-# LANGUAGE NamedFieldPuns #-}
{-# LANGUAGE DisambiguateRecordFields #-}
{-# LANGUAGE RecordWildCards #-}
{-# LANGUAGE FlexibleContexts #-}

module BoardGame.Server.Domain.Board (
    Board
  , Board(Board)
  , Board(dimension)
  , Board(grid)
  , mkBoard
  , mkEmptyBoard
  , mkBoardFromPieces
  , rowsAsPieces
  , rowsAsStrings
  , colsAsPieces
  , next
  , prev
  , adjacent
  , get
  , getGridPieces
  , set
  , setN
  , setDeadPoints
  , isEmpty
  , stripOfPlay
  , inBounds
  , pointIsEmpty
  , pointIsNonEmpty
  , pointHasValue
  -- , pointIsIsolatedInLine
  , pointHasRealNeighbor
  , validateCoordinate
  , validatePoint
  -- , farthestNeighbor
  -- , surroundingRange
  , getLetter
  , stripIsDisconnectedInLine
  , playableEnclosingStripsOfBlankPoints
  , computeAllLiveStrips
  -- , groupedStrips
  , lineNeighbors
)
where

import Data.List
import qualified Data.Maybe as Maybe
import Data.Map (Map)
import qualified Data.Map as Map
import Data.Maybe (fromJust, isJust)

-- import qualified Data.ByteString.Char8 as BS

import Control.Monad.Except (MonadError(..))

import BoardGame.Common.Domain.PlayPiece (PlayPiece, PlayPiece(PlayPiece))
import qualified BoardGame.Common.Domain.PlayPiece as PlayPiece
import Bolour.Plane.Domain.GridValue (GridValue(GridValue))
import BoardGame.Common.Domain.GridPiece (GridPiece)
import qualified BoardGame.Common.Domain.GridPiece as GridPiece
import qualified Bolour.Plane.Domain.GridValue as GridValue
import BoardGame.Common.Domain.Piece (Piece, Piece(Piece))
import qualified BoardGame.Common.Domain.Piece as Piece
import Bolour.Plane.Domain.Axis (Coordinate, Axis(..))
import Bolour.Plane.Domain.Point (Point, Point(Point))
import qualified Bolour.Plane.Domain.Point as Point
import qualified Bolour.Plane.Domain.Axis as Axis
import Bolour.Util.BlackWhite (BlackWhite, BlackWhite(Black, White))
import qualified Bolour.Util.BlackWhite as BlackWhite
import Bolour.Plane.Domain.BlackWhitePoint (BlackWhitePoint, BlackWhitePoint(BlackWhitePoint))
import qualified Bolour.Plane.Domain.BlackWhitePoint as BlackWhitePoint
import Bolour.Plane.Domain.BlackWhiteGrid (BlackWhiteGrid)
import qualified Bolour.Plane.Domain.BlackWhiteGrid as Gr
import BoardGame.Server.Domain.GameError (GameError(..))
import qualified Bolour.Util.MiscUtil as Util
import BoardGame.Server.Domain.Strip (Strip, Strip(Strip))
import qualified BoardGame.Server.Domain.Strip as Strip
import qualified Bolour.Util.Empty as Empty

{--
   A Board uses a BlackWhiteGrid to represent the contents of the board.
   Within the grid, an empty slot is represented
   by Maybe.Nothing. But Board exposes pieces that represent
   emptiness by their value being null: '\0'. Where necessary,
   this module translates between these two representations of
   emptiness.
--}

-- | The game board.
data Board = Board {
    dimension :: Int
  , grid :: BlackWhiteGrid Piece
}
  deriving (Show)

pieceToBlackWhite :: Piece -> BlackWhite Piece
pieceToBlackWhite piece | piece == Piece.deadPiece = Black
                        | piece == Piece.emptyPiece = White Nothing
                        | otherwise = White (Just piece)

blackWhiteToPiece :: BlackWhite Piece -> Piece
blackWhiteToPiece Black = Piece.deadPiece
blackWhiteToPiece (White Nothing) = Piece.emptyPiece
blackWhiteToPiece (White (Just piece)) = piece

pieceExtractor :: BlackWhitePoint Piece -> Piece
pieceExtractor BlackWhitePoint {value = bwPiece, point} = blackWhiteToPiece bwPiece

-- TODO. Check rectangular. Check parameters. See below.
mkBoardFromPieces :: [[Maybe Piece]] -> Int -> Board
mkBoardFromPieces cells =
  let cellMaker row col = White $ cells !! row !! col
  in mkBoard' cellMaker

-- TODO. Ditto.
mkBoard :: (Int -> Int -> Piece) -> Int -> Board
mkBoard pieceMaker =
  let cellMaker row col = White $ Piece.toMaybe $ pieceMaker row col
  in mkBoard' cellMaker

mkEmptyBoard :: Int -> Board
mkEmptyBoard dimension =
  let grid = Gr.mkEmptyGrid dimension dimension
  in Board dimension grid

mkBoard' :: (Int -> Int -> BlackWhite Piece) -> Int -> Board
mkBoard' cellMaker dimension =
  let grid = Gr.mkGrid cellMaker dimension dimension
  in Board dimension grid

rowsAsPieces :: Board -> [[Piece]]
rowsAsPieces Board {grid} =
  let lineMapper row = pieceExtractor <$> row
  in lineMapper <$> Gr.rows grid

colsAsPieces :: Board -> [[Piece]]
colsAsPieces Board {grid} =
  let lineMapper col = pieceExtractor <$> col
  in lineMapper <$> Gr.cols grid

next :: Board -> Point -> Axis -> Maybe (BlackWhite Piece)
next Board {grid} point axis = BlackWhitePoint.value <$> Gr.next grid point axis

prev :: Board -> Point -> Axis -> Maybe (BlackWhite Piece)
prev Board {grid} point axis = BlackWhitePoint.value <$> Gr.prev grid point axis

adjacent :: Board -> Point -> Axis -> Int -> Maybe (BlackWhite Piece)
adjacent Board {grid} point axis direction = BlackWhitePoint.value <$> Gr.adjacent grid point axis direction

-- | Nothing if out of bounds, noPiece if empty but in bounds.
get :: Board -> Point -> Maybe Piece
get board @ Board { grid } point =
  if not (inBounds board point) then Nothing
  else
    let bwVal = Gr.get grid point
    in Just $ blackWhiteToPiece bwVal

-- | Assume point is valid.
getLetter :: Board -> Point -> Char
getLetter board point =
  Piece.value $ Maybe.fromJust $ get board point

getGridPieces :: Board -> [GridPiece]
getGridPieces Board {grid} =
  let locatedPieces = Gr.getValues grid
      toGridPiece (piece, point) = GridValue piece point
  in toGridPiece <$> locatedPieces

set :: Board -> Point -> Piece -> Board
set Board { dimension, grid } point piece =
  let bwPiece = pieceToBlackWhite piece
      grid' = Gr.set grid point bwPiece
  in Board dimension grid'

setN :: Board -> [GridPiece] -> Board
setN board @ Board {dimension, grid} gridPoints =
  let toBWPoint GridValue {value = piece, point} = BlackWhitePoint (pieceToBlackWhite piece) point
      bwPoints = toBWPoint <$> gridPoints
      grid' = Gr.setN grid bwPoints
  in Board dimension grid'

setDeadPoints :: Board -> [Point] -> Board
setDeadPoints board points =
  let deadGridPiece point = GridValue Piece.deadPiece point
      deadGridPieces = deadGridPiece <$> points
  in setN board deadGridPieces

isEmpty :: Board -> Bool
isEmpty Board { grid } = Empty.isEmpty grid
  -- let cellList = concat $ Gr.rows grid
  -- in all (Maybe.isNothing . fst) cellList

pointIsEmpty :: Board -> Point -> Bool
pointIsEmpty Board {grid} point = Gr.isEmpty grid point
  -- Maybe.isNothing $ Gr.get grid point

pointIsNonEmpty :: Board -> Point -> Bool
pointIsNonEmpty board point = not $ pointIsEmpty board point

pointHasValue :: Board -> Point -> Bool
pointHasValue Board {grid} point = Gr.hasValue grid point

inBounds :: Board -> Point -> Bool
inBounds Board {grid} = Gr.inBounds grid

validateCoordinate :: MonadError GameError m =>
  Board -> Axis -> Coordinate -> m Coordinate

validateCoordinate (board @ Board { dimension }) axis coordinate =
  if coordinate >= 0 && coordinate < dimension then return coordinate
  else throwError $ PositionOutOfBoundsError axis (0, dimension) coordinate

validatePoint :: MonadError GameError m =>
  Board -> Point -> m Point

validatePoint board (point @ Point { row, col }) = do
  _ <- validateCoordinate board Y row
  _ <- validateCoordinate board X col
  return point

rowsAsStrings :: Board -> [String]
rowsAsStrings board = ((\Piece {value} -> value) <$>) <$> rowsAsPieces board

maybeBlackWhiteHasPiece :: Maybe (BlackWhite Piece) -> Bool
maybeBlackWhiteHasPiece Nothing = False
maybeBlackWhiteHasPiece (Just bwPiece) = BlackWhite.hasValue bwPiece

pointHasRealNeighbor :: Board -> Point -> Axis -> Bool
pointHasRealNeighbor board point axis =
  let maybeNext = next board point axis
      maybePrev = prev board point axis
  in maybeBlackWhiteHasPiece maybeNext || maybeBlackWhiteHasPiece maybePrev

stripOfPlay :: Board -> [PlayPiece] -> Maybe Strip
stripOfPlay board [] = Nothing
stripOfPlay board [playPiece] = stripOfPlay1 board playPiece
stripOfPlay board (pp:pps) = stripOfPlayN board (pp:pps)

stripOfPlay1 :: Board -> PlayPiece -> Maybe Strip
stripOfPlay1 board PlayPiece {point} =
  let Point {row, col} = point
      -- Arbitrarily choose the row of the single move play.
      line = rowsAsPieces board !! row
      lineAsString = Piece.piecesToString line
  in Just $ Strip.lineStrip Axis.X row lineAsString col 1

stripOfPlayN :: Board -> [PlayPiece] -> Maybe Strip
stripOfPlayN board playPieces =
  let rows = rowsAsPieces board
      cols = colsAsPieces board
      points = (\PlayPiece {point} -> point) <$> playPieces
      Point {row = hdRow, col = hdCol} = head points
      maybeAxis = Point.axisOfLine points
  in
    let mkStrip axis =
          let (lineNumber, line, begin) =
                case axis of
                  Axis.X -> (hdRow, rows !! hdRow, hdCol)
                  Axis.Y -> (hdCol, cols !! hdCol, hdRow)
              lineAsString = Piece.piecesToString line
           in Strip.lineStrip axis lineNumber lineAsString begin (length points)
    in mkStrip <$> maybeAxis

-- surroundingRange :: Board -> Point -> Axis -> [Point]
-- surroundingRange Board {grid} = Gr.surroundingRange grid

-- | Check that a strip has no neighbors on either side - is disconnected
--   from the rest of its line. If it is has neighbors, it is not playable
--   since a matching word will run into the neighbors. However, a containing
--   strip will be playable and so we can forget about this strip.
stripIsDisconnectedInLine :: Board -> Strip -> Bool
stripIsDisconnectedInLine board (strip @ Strip {axis, begin, end, content})
  | null content = False
  | otherwise =
      let f = Strip.stripPoint strip 0
          l = Strip.stripPoint strip (end - begin)
          -- limit = dimension
          maybePrevPiece = prev board f axis
          maybeNextPiece = next board l axis
          isSeparator maybeBlackWhitePiece = not $ maybeBlackWhiteHasPiece maybeBlackWhitePiece
      in isSeparator maybePrevPiece && isSeparator maybeNextPiece

computeAllLiveStripsForAxis :: Board -> Axis -> [Strip]
computeAllLiveStripsForAxis board axis =
  let lines = case axis of
                Axis.X -> rowsAsPieces board
                Axis.Y -> colsAsPieces board
  in Strip.allLiveStrips axis (Piece.piecesToString <$> lines)

computeAllLiveStrips :: Board -> [Strip]
computeAllLiveStrips board =
  computeAllLiveStripsForAxis board Axis.X ++ computeAllLiveStripsForAxis board Axis.Y

enclosingStripsOfBlankPoints :: Board -> Axis -> Map.Map Point [Strip]
enclosingStripsOfBlankPoints board axis =
  let liveStrips = computeAllLiveStripsForAxis board axis
      stripsEnclosingBlanks = filter Strip.hasBlanks liveStrips
  in Util.inverseMultiValuedMapping Strip.blankPoints stripsEnclosingBlanks

playableEnclosingStripsOfBlankPoints :: Board -> Axis -> Int -> Map.Map Point [Strip]
playableEnclosingStripsOfBlankPoints board axis trayCapacity =
  let enclosing = enclosingStripsOfBlankPoints board axis
      playable strip @ Strip {blanks, content} =
        blanks <= trayCapacity &&
          stripIsDisconnectedInLine board strip &&
          length content > 1 -- Can't play to a single blank strip - would have no anchor.
  in
    filter playable <$> enclosing

-- | Get all the colinear neighbors in a given direction along a given axis
--   ordered in increasing value of the line index (excluding the point
--   itself). A colinear point is considered a neighbor if it has a real value,
--   and is adjacent to the given point, or recursively adjacent to a neighbor.
lineNeighbors :: Board -> Point -> Axis -> Int -> [GridPiece]
lineNeighbors board @ Board {grid} point axis direction =
  let piecePointPairs = Gr.lineNeighbors grid point axis direction
  in toGridPiece <$> piecePointPairs
     where toGridPiece (piece, point) = GridValue piece point

File: scala-server/app/com/bolour/boardgame/scala/server/domain/Board.scala
/*
 * Copyright 2017 Azad Bolour
 * Licensed under GNU Affero General Public License v3.0 -
 *    https://github.com/azadbolour/boardgame/blob/master/LICENSE.md
 */
package com.bolour.boardgame.scala.server.domain

import com.bolour.plane.scala.domain.Axis.Axis
import com.bolour.boardgame.scala.common.domain._
import com.bolour.boardgame.scala.server.domain.GameExceptions.InternalGameException
import com.bolour.plane.scala.domain._
import com.bolour.util.scala.common.CommonUtil.inverseMultiValuedMapping
import com.bolour.util.scala.common.{Black, BlackWhite, White}

case class Board(dimension: Int, grid: BlackWhiteGrid[Piece]) {

  import Board._

  /**
    * Is the board empty? Empty means all of its points are White(None)?
    * Note that a black point makes the board non-empty in this
    * interpretation.
    */
  def isEmpty: Boolean = grid.isEmpty

  /**
    * If there is a piece at the given point return it in Some, otherwise
    * return None.
    */
  def getPiece(point: Point): Option[Piece] = {
    val bw = grid.get(point)
    bw match {
      case White(Some(piece)) => Some(piece)
      case _ => None
    }
  }

  /**
    * Get the value of a point on the board (if the point is out of bounds return black -
    * same as if the point is black).
    */
  def get(point: Point): BlackWhite[Piece] = grid.get(point)

  /**
    * Get the board pieces and their locations as grid pieces.
    */
  def gridPieces: List[GridPiece] = {
    val piecesAndPoints = grid.getValues
    piecesAndPoints map { case (piece, point) => GridPiece(piece, point)}
  }

  /**
    * Update a number of values of the board returning the updated board.
    * @param bwPoints List of (piece, point) pairs (housed in BlackWhitePoints) to update.
    *                 Out of bound points are ignored.
    */
  def setN(bwPoints: List[BlackWhitePoint[Piece]]): Board = {
    val augmentedGrid = grid.setN(bwPoints)
    Board(dimension, augmentedGrid)
  }

  /**
    * Similar to setN but provides the (piece, point) pairs in a list of GridPieces.
    */
  def setGridPieces(gridPieces: List[GridPiece]): Board = {
    val bwPoints = gridPieces map { case GridPiece(piece, point) => BlackWhitePoint(White(Some(piece)), point)}
    setN(bwPoints)
  }

  /**
    * Set a number of the board's points to black.
    */
  def setBlackPoints(deadPoints: List[Point]): Board = {
    val bwPoints = deadPoints map { point => BlackWhitePoint(Black[Piece](), point) }
    setN(bwPoints)
  }

  private def rows = grid.rows
  private def columns = grid.columns

  private def lineToString(bwPoints: List[BlackWhitePoint[Piece]]): String = {
    val chars = bwPoints map { case BlackWhitePoint(bwPiece, _) => Piece.bwPieceToChar(bwPiece) }
    chars.mkString
  }

  /**
    * Given a play in terms of a list of play pieces, find the strip
    * on which the play is made.
    *
    * TODO. Assumes play is contiguous.
    */
  def stripOfPlay(playPieces: List[PlayPiece]): Strip = {
    val len = playPieces.length
    if (len == 0)
      throw new InternalGameException("no pieces in play", null)

    if (len == 1) {
      val PlayPiece(_, point, _) = playPieces.head
      // Arbitrarily consider the single play it as a horizontal play.
      val theRow = rows(point.row)
      val content = lineToString(theRow)
      return Strip.lineStrip(Axis.X, point.row, content, point.col, point.col)
    }

    val points = playPieces.map(_.point)

    val head = points.head
    val next = points(1)

    val axis = if (head.row == next.row) Axis.X else Axis.Y
    val (lineNumber, line, begin) = axis match {
      case Axis.X => (head.row, rows(head.row), head.col)
      case Axis.Y => (head.col, columns(head.col), head.row)
    }
    val end = begin + points.length - 1
    val content = lineToString(line)
    Strip.lineStrip(axis, lineNumber, content, begin, end)
  }

  def pointIsEmpty(point: Point): Boolean = grid.get(point).isEmpty

  def inBounds(point: Point): Boolean = {
    val Point(row, col) = point
    inBounds(row) && inBounds(col)
  }

  def inBounds(coordinate: Int): Boolean = coordinate >= 0 && coordinate < dimension

  def colinearPoint(point: Point, axis: Axis, direction: Int)(steps: Int): Option[Point] = {
    val nth = point.colinearPoint(axis, direction)(steps)
    if (!inBounds(nth)) None else Some(nth)
  }

  /**
    * Considering a point to be a neighbor if it is recursively adjacent to
    * the given point in the given direction, find the farthest one in that direction.
    */
  def farthestNeighbor1(point: Point, axis: Axis, direction: Int): Point = {

    def outOfBoundsOrEmpty(oPoint: Option[Point]): Boolean =
      oPoint.isEmpty || pointIsEmpty(oPoint.get)

    def adjacent(p: Point): Option[Point] = colinearPoint(p, axis, direction)(1)

    def isBoundary(p: Point): Boolean =
      !pointIsEmpty(p) && outOfBoundsOrEmpty(adjacent(p))

    // The starting point is special because it is empty.
    if (outOfBoundsOrEmpty(adjacent(point)))
      return point

    val neighbors = (1 until dimension).toList map colinearPoint(point, axis, direction)
    val farthest = neighbors find (_.exists(isBoundary))
    farthest.get.get // A boundary always exists.
  }

  def farthestNeighbor(point: Point, axis: Axis, direction: Int): Point = {

    def isBoundary(p: Point): Boolean =
      getPiece(p).isDefined && getPiece(p.adjPoint(axis, direction)).isEmpty

    // The starting point is special because it is empty.
    if (getPiece(point.adjPoint(axis, direction)).isEmpty)
      return point

    val colinears = (1 until dimension).toList map colinearPoint(point, axis, direction)
    val farthest = colinears find (_.exists(isBoundary))
    farthest.get.get // A boundary always exists.
  }

  /**
    * Get the value of the next point on the grid (None if next is off the grid).
    */
  def next(point: Point, axis: Axis): Option[BlackWhite[Piece]] =
    grid.next(point, axis) map { _.value }

  /**
    * Get the value of the previous point on the grid (None if previous is off the grid).
    */
  def prev(point: Point, axis: Axis): Option[BlackWhite[Piece]] =
    grid.prev(point, axis) map { _.value }

  /**
    * Get the value of an adjacent point on the grid (None if adjacent is off the grid).
    */
  def adjacent(point: Point, axis: Axis, direction: Int): Option[BlackWhite[Piece]] =
    grid.adjacent(point, axis, direction) map { _.value }


  def hasRealNeighbor(point: Point, axis: Axis): Boolean = {
    val nextOpt = next(point, axis)
    if (nextOpt.isDefined && nextOpt.get.hasValue)
      return true
    val prevOpt = prev(point, axis)
    if (prevOpt.isDefined && prevOpt.get.hasValue)
      return true
    false
  }

  def playableEmptyStrips(traySize: Int): List[Strip] = {
    val center = dimension/2
    val strips = grid.segmentsForLineNumber(Axis.X, center) map lineSegmentToStrip
    val conformantStrips = strips.filter { strip => strip.begin <= center && strip.end >= center}
    conformantStrips
  }

  def playableStrips(traySize: Int): List[Strip] = {
    val allStrips = computeAllLiveStrips
    def hasFillableBlanks = (s: Strip) => s.numBlanks > 0 && s.numBlanks <= traySize
    val conformantStrips1 = allStrips.filter(hasFillableBlanks)
    val conformantStrips2 = conformantStrips1.filter(_.hasAnchor)
    val conformantStrips3 = conformantStrips2.filter(stripIsDisconnectedInLine)
    conformantStrips3
  }

  def potentialPlayableStrips(axis: Axis, trayCapacity: Int): List[Strip] = {
    val allStrips = grid.segmentsAlongAxis(axis) map lineSegmentToStrip
    def hasFillableBlanks = (s: Strip) => s.numBlanks > 0 && s.numBlanks <= trayCapacity
    val conformantStrips1 = allStrips.filter(hasFillableBlanks)
    val conformantStrips2 = conformantStrips1.filter(stripIsDisconnectedInLine)
    conformantStrips2
  }

  def stripIsDisconnectedInLine(strip: Strip): Boolean = {
    val firstPoint = strip.point(0)
    val lastPoint = strip.point(strip.end - strip.begin)
    val maybePrevPiece = prev(firstPoint, strip.axis)
    val maybeNextPiece = next(lastPoint, strip.axis)

    def isSeparator(maybeBlackWhitePiece: Option[BlackWhite[Piece]]): Boolean = {
      maybeBlackWhitePiece match {
        case None => true
        case Some(bwPiece) =>
          bwPiece match {
            case Black() => true
            case White(None) => true
            case White(Some(_)) => false
          }
      }
    }
    isSeparator(maybePrevPiece) && isSeparator(maybeNextPiece)
  }

  def computeAllLiveStrips : List[Strip] = grid.allSegments map lineSegmentToStrip

  def enclosingStripsOfBlankPoints(axis: Axis): Map[Point, List[Strip]] = {
    val stripsEnclosingBlanks = (grid.segmentsAlongAxis(axis) map lineSegmentToStrip) filter { _.numBlanks > 0 }
    inverseMultiValuedMapping((strip: Strip) => strip.blankPoints)(stripsEnclosingBlanks)
  }

  def playableEnclosingStripsOfBlankPoints(axis: Axis, trayCapacity: Int): Map[Point, List[Strip]] = {
    val enclosing = enclosingStripsOfBlankPoints(axis)
    val playable = enclosing mapValues { (strips: List[Strip]) =>
      strips filter { (s: Strip) =>
        s.numBlanks <= trayCapacity &&
          stripIsDisconnectedInLine(s) &&
          // Can't play to a single blank strip - would have no anchor.
          s.content.length > 1
      }
    }
    playable
  }
}

object Board {
  def apply(dimension: Int, cellMaker: Int => Int => BlackWhite[Piece]) : Board = {
    val grid = BlackWhiteGrid[Piece](cellMaker, dimension, dimension)
    Board(dimension, grid)
  }

  def apply(dimension: Int) : Board = {
    def cellMaker(row: Int)(col: Int): BlackWhite[Piece] = White(None)
    val grid = BlackWhiteGrid[Piece](cellMaker _, dimension, dimension)
    Board(dimension, grid)
  }

  /**
    * Create a board with the given pieces at the given positions in the grid pieces.
    */
  def apply(dimension: Int, gridPieces: List[GridPiece]): Board = {
    def maybeGridPiece(r: Int, c: Int) = gridPieces.find(_.point == Point(r, c))
    def cellMaker(row: Int)(col: Int): BlackWhite[Piece] = {
      maybeGridPiece(row, col) match {
        case Some(GridPiece(piece, point)) => White(Some(piece))
        case None => White(None)
      }
    }
    Board(dimension, cellMaker _)
  }

  def lineSegmentToStrip(lineSegment: LineSegment[Piece]): Strip = {
    def optToChar(opt: Option[Piece]): Char =
      opt match {
        case None => ' '
        case Some(piece) => piece.value
      }

    lineSegment match {
      case LineSegment(axis, lineNumber, begin, end, segment) =>
        Strip(axis, lineNumber, begin, end, (segment map optToChar).mkString)
    }
  }
}

